### ISSUE TITLE
Camera position optimization using differentiable rendering didn't work well

### ISSUE THREAD

**Issue by Haitham96** [facebookresearch/pytorch3d#127]
I have a obj 3d for a man and i have the silhouete of his picture from the front i want get the camera position that match the silhouete.
so i followed your amazing implementation in [Camera position optimization using differentiable rendering](https://github.com/facebookresearch/pytorch3d/blob/master/docs/tutorials/camera_position_optimization_with_differentiable_rendering.ipynb).
This is the starting position:
![image](https://user-images.githubusercontent.com/45899818/77693851-b1235a00-6fb1-11ea-8841-c14f1c32a8cc.png)

After that i run the optimizer like the tutorial but i got a bad result the loss didn't change

![image](https://user-images.githubusercontent.com/45899818/77694451-b3d27f00-6fb2-11ea-9792-066c31691d97.png)
![image](https://user-images.githubusercontent.com/45899818/77694503-c6e54f00-6fb2-11ea-88c0-582f47fe08df.png)

How i can get a better result?

**Comment by nikhilaravi** [comment]
@Haitham96 to debug, can you try changing the initial position of the camera (inside `Model` in the notebook) so the camera is a bit closer to the object?  The starting silhouette looks a bit strange..  Are you using a reference image loaded from an existing file or one that is generated following the steps in the tutorial?

**Comment by Haitham96** [comment]
yes i using a reference image loaded from an existing file.
i changed the initial camera position to [1.0,  1.0, 1.0]
![image](https://user-images.githubusercontent.com/45899818/77697789-2abe4680-6fb8-11ea-962a-caa3b64c422e.png)


but in the optimization step i got the same

![image](https://user-images.githubusercontent.com/45899818/77697838-41649d80-6fb8-11ea-9286-ec1bfccbbd04.png)


![image](https://user-images.githubusercontent.com/45899818/77697859-4de8f600-6fb8-11ea-9335-67b600be3957.png)

**Comment by Haitham96** [comment]
Hi @nikhilaravi is there any idea how i can get it works?

**Comment by PascalYIN** [comment]
Hello there I have met a similar problem while trying to align 2 images using differentiable rendering, and I found something interesting, I would like to share my experience with you, and hopefully it might help.

I was using a dim6 vector of se3 to represent a camera position, which is transformed as a 4*4 matrix by se3 exponential mapping, noted M (shape 4*4). so R = M[:3, :3],  and T=M[3, :3] by your convention. I then put R and T in the renderer's setting. I used a summed least-square as loss function as shown in your l [tutoria](https://pytorch3d.org/tutorials/camera_position_optimization_with_differentiable_rendering)

The interesting part is when I started the optimization the loss augmented smoothly, totally against my expectation, and it just kept augmenting. So by curiosity I manually changed the grad value of my model's parameters and then the loss began to fall, as smoothly as it augmented during the first time. So this makes me think there might be some sort of sign error during the computation of gradient ?

**Comment by Haitham96** [comment]
Hi @PascalYIN thank you for sharing your experience, How changed them manually?

**Comment by Haitham96** [comment]
**update:** i tired to print R and T and i got a tensors of non values/
![image](https://user-images.githubusercontent.com/45899818/77854149-26c93900-71f1-11ea-8750-1db61ad6a951.png)

**Comment by Haitham96** [comment]
@nikhilaravi i also tried to use a  reference image just like that is generated following the steps in the tutorial, i didn't change any thing in the tutorial just upload my object instead of yours and change this parameters:
`distance = 2   # distance from camera to the object
elevation = 20.0   # angle of elevation in degrees
azimuth = 0.0  # No rotation so the camera is positioned on the +Z axis. `
and then the loss didn't change in the optimization

**Comment by PascalYIN** [comment]
@Haitham96 Hi, for example x is the parameter that I what to calculate its gradient. I manually set 
`x.grad = -x.grad` and that's it.

But I have impression that the optimization works by chance, because the loss in my experience dropped only for several frames (with the grad set to be negative), and for other frames the loss always augmented either for +1 * grad or -1 * grad

**Comment by gkioxari** [comment]
We have fixed the tutorial which had a small issue regarding the background color. Closing this. Reopen if you have more questions.
